A simple kill switch implementation.

alxmrs opened the pull request:
Adding a `target_epochs` argument to all LR schedulers that allows us to simulate long LR schedules (that actually stop short).

alxmrs committed:
A simple kill switch implementation.

Added a simple integration test as well. I am open to other naming suggestions for the config.

jder commented:
I was thinking more about this -- is it really just about LR schedule vs run length? If so maybe we add an optional `target_epochs` to the cosine scheduler config instead?

alxmrs commented:
Sorry for the late reply. I'm not sure how this would fit into the lr scheduler. Can you explain more? 

My intention is to not interrupt with the normal LR schedule while also ending the training runs early in order to iterate through many configurations.

jder commented:
I'm saying that the epochs config is already a "kill after N epochs" switch. Rather than adding a second one I'm suggesting you instead add a config option to the lr scheduler which tells it how long the decay should be over. 

So rather than:

epochs = 70
kill_after_n_epochs = 5

You'd have:

epochs = 5
lr_scheduler = {type: cosine, decay_epochs: 70}

(Where decay_epochs = None means the current behavior)

alxmrs commented:
Thanks for the explanation -- I now understand your suggestion. I'm happy to implement that! Now that I've seen how to implement a new lr scheduler (#343), I see how I could implement this.

alxmrs committed:
Revert "A simple kill switch implementation."

This reverts commit b809d47795b66d5942380249b08edd8d4ee34569.

alxmrs committed:
"Kill switch" based on LR scheduler config.

It's an open question if the tolerances are acceptable.

# Conflicts:
#	tests/test_schedule.py

alxmrs commented on tests/test_schedule.py:
```diff
@@ -49,3 +57,44 @@ def test_cosine_with_tail_holds_constant_lr():
+        # TODO(alxmrs): Is this an OK tolerance?
+        assert abs(short_lr - base_lr) < 1e-3, (
+            "Shortened schedule should be the same shape as standard LR schedule"
+        )
```
@jder It seems like this alternative method does not produce exactly the same schedule -- this tolerance looks pretty coarse to me. WDYT about going with this implementation vs the original, given this?

jder replied:
It seemed logically impossible to me that this method would not produce the same schedule (since the schedule has no way of knowing how long we're gonna run it for), so I grabbed the code and did some digging. The problem is that the optimizer was being shared between the baseline and test schedule simulations, and so it maintained some state between those two. Pushed up a fix, probably also needs to be applied to the other test.

alxmrs replied:
ah, that's surprising, thanks for digging into this! will TAL now.

alxmrs replied:
Thanks, this is a great contribution. I took the liberty of pulling the repeated optimizer setup into the "simulate" function.

jder committed:
fix bug in test_cosine__larger_target_than_total__stops_early

alxmrs committed:
rm TODO (thanks for the fix, @jder).

alxmrs committed:
Extracted optimizer into helper fn.

alxmrs committed:
Added a test that the curve is, indeed, a cosine schedule.

jder reviewed (approved):
Thanks! Sorry, thought I was still waiting for you on this.

alxmrs committed:
Fix test (from hard merge issue).

alxmrs committed:
Adding total epochs to the CosineWithWarmup scheduler.

alxmrs commented:
No problem; it wasn't blocking anything.
